pkg/csi/service/common/commonco/k8sorchestrator/topology.go

```diff
@@ -1620,8 +1621,72 @@ func (volTopology *wcpControllerVolumeTopology) GetTopologyInfoFromNodes(ctx con
 
 	switch strings.ToLower(params.StorageTopologyType) {
 	case "zonal":
-		// If the topology requirement received has just one zone, use the same zone as node affinity terms on PV.
-		if len(params.TopologyRequirement.GetPreferred()) == 1 {
+		if params.TopologyRequirement == nil {
+			// This case is for static volume provisioning using CNSRegisterVolume API
+			if !isPodVMOnStretchedSupervisorEnabled {
+				return nil, logger.LogNewErrorf(log, "topology requirement should not be nil. invalid params: %v", params)
+			} else {
+				// If the topology requirement received is nil, then identify topology of the datastore by looking into
+				// azClustersMap
+				var selectedSegments []map[string]string
+				for az, clusters := range azClustersMap {
+					var sharedDatastoresForclusterMorefs []*cnsvsphere.DatastoreInfo
+					for index, clusterMoref := range clusters {
+						accessibleDs, _, err := cnsvsphere.GetCandidateDatastoresInCluster(ctx, params.Vc, clusterMoref)
+						if err != nil {
+							return nil, logger.LogNewErrorf(log,
+								"failed to find candidate datastores to place volume in cluster %q. Error: %v",
+								clusterMoref, err)
+						}
+						if len(accessibleDs) == 0 {
+							return nil, logger.LogNewErrorf(log,
+								"no shared candidate datastores found to place volume for clusters %v", clusters)
```

Comment:
shall we specify particular cluster moref in the log message, instead of clusters?

Reply:
Yes here it we print specific cluster which is not returning accessibleDs it would help.

Reply:
done.